File: 프로그래머스/1/389478. 택배 상자 꺼내기/택배 상자 꺼내기.py
def get_box_counts(w, box, floor):
    left = box - (w * floor)
    right = w * (floor + 1) - (box + 1)
    
    if floor % 2 == 1:
        left, right = right, left
        
    return left, right

def get_next_box(box, left, right, floor):
    if floor % 2 == 0:
        return box + (2 * right + 1)
    else:
        return box + (2 * left + 1)


def solution(n, w, num):
    answer = 1
    n, box = n - 1, num - 1
    
    floor = box // w
    left, right = get_box_counts(w, box, floor)
    
    while True:
        next_box = get_next_box(box, left, right, floor)
        if next_box > n:
            break
        
        box = next_box
        answer += 1
        floor += 1
        
    return answer

File: 프로그래머스/1/389478. 택배 상자 꺼내기/택배 상자 꺼내기.java
class Solution {
    class BoxInfo {
        public int left;
        public int right;
        
        public BoxInfo(int left, int right) {
            this.left = left;
            this.right = right;
        }
    }
    
    private BoxInfo getBoxInfo(int w, int box, int floor) {
        int left = box - (w * floor);
        int right = w * (floor + 1) - (box + 1);
        
        if (floor % 2 == 1) {
            int tmp = left;
            left = right;
            right = tmp;
        }
        
        return new BoxInfo(left, right);
    }
    
    private int getNextBox(int box, int left, int right, int floor) {
        if (floor % 2 == 0) {
            return box + (2 * right + 1);
        } else {
            return box + (2 * left + 1);
        }
    }
    
    public int solution(int n, int w, int num) {
        int answer = 1;
        int box = num - 1;
        n--;
        
        int floor = box / w;
        BoxInfo info = getBoxInfo(w, box, floor);
        
        int left = info.left;
        int right = info.right;
        
        while (true) {
            int nextBox = getNextBox(box, left, right, floor);
            if (nextBox > n) {
                break;
            }
            
            box = nextBox;
            answer += 1;
            floor += 1;
        }
        
        return answer;
    }
}